job/server/src/main/java/alluxio/master/job/JobMaster.java

```diff
@@ -162,35 +187,40 @@ public String getName() {
    */
   public synchronized long run(JobConfig jobConfig)
       throws JobDoesNotExistException, ResourceExhaustedException {
-    long jobId = mJobIdGenerator.getNewJobId();
-    JobInfo jobInfo = new JobInfo(jobId, jobConfig, new Function<JobInfo, Void>() {
-      @Override
-      public Void apply(JobInfo jobInfo) {
-        Status status = jobInfo.getStatus();
-        mFinishedJobs.remove(jobInfo);
-        if (status.isFinished()) {
-          mFinishedJobs.add(jobInfo);
-        }
-        return null;
-      }
-    });
     if (mIdToJobCoordinator.size() == CAPACITY) {
       if (mFinishedJobs.isEmpty()) {
         // The job master is at full capacity and no job has finished.
         throw new ResourceExhaustedException("Job master is at full capacity");
       }
-      // Check if the oldest finished job can be discarded.
+      // Discard old jobs that have completion time beyond retention policy
       Iterator<JobInfo> jobIterator = mFinishedJobs.iterator();
```

Comment:
It was not. Changed the list implementation to ConcurrentSkipListSet, which is a SortedSet which is concurrent.

Reply:
FYI, weak consistency of a concurrent collection is acceptable here. Recently changed elements are of no concern anyway because of the retention policy.